syntax = "proto3";

// Namespace
package hive;

// Execute
message TaskRequest {
    // Task ID
    bytes task_id = 1;

    // Future ID
    bytes future_id = 2;

    // Function
    bytes function = 3;

    // Args
    bytes args = 4;

    // Keyword Arguments
    bytes kwargs = 5;

    // Object IDs
    bytes object_locs = 6;
}

message TaskReply {
    // Task ID
    bytes task_id = 1;

    // Return Value
    bytes result = 3;

    // Object IDs {"missing" : [Missing Object IDs], "current" : [Current Object IDs]}
    // 2 lists, index 0: missing objects, index 1: current objects
    bytes object_ids = 2;
}

// GetLoad
message LoadRequest {
}

message LoadReply {
    // CPU Load
    bytes cpu_load = 1;

    // Memory Utilization
    bytes memory_used = 2;
}

// Driver to Worker
service DriverWorkerService {
    rpc Execute (TaskRequest) returns (TaskReply);
    rpc GetLoad (LoadRequest) returns (LoadReply);
}